File: day22/day22-2.cpp
#include <algorithm>
#include <fstream>
#include <numeric>
#include <vector>
#include <fmt/format.h>

using namespace std;

struct Vec3
{
	int v[3];
};

struct Cuboid
{
	Vec3 p[2];

	bool inside(const Cuboid& o) const
	{
		for (int i = 0; i < 3; i++)
		{
			if (p[0].v[i] < o.p[0].v[i] || p[1].v[i] > o.p[1].v[i])
			{
				return false;
			}
		}
		return true;
	}

	bool intersect(const Cuboid& o) const
	{
		for (int i = 0; i < 3; i++)
		{
			if (p[1].v[i] <= o.p[0].v[i] || p[0].v[i] >= o.p[1].v[i])
			{
				return false;
			}
		}
		return true;
	}

	uint64_t volume() const
	{
		uint64_t v = 1;
		for (int i = 0; i < 3; i++)
		{
			v *= p[1].v[i] - p[0].v[i];
		}
		return v;
	}
};

struct Instr
{
	enum { ON, OFF } cmd;
	Cuboid c;
};

istream& operator>>(istream& in, vector<Instr>& ins)
{
	ins.clear();
	while (in)
	{
		Instr tmp{};
		string cmd;
		in >> cmd;
		if (cmd == "on")
		{
			tmp.cmd = Instr::ON;
		}
		else if (cmd == "off")
		{
			tmp.cmd = Instr::OFF;
		}
		else
		{
			break;
		}

		for (int i = 0; i < 3; i++)
		{
			in.seekg(3, ios::cur);
			in >> tmp.c.p[0].v[i];
			in.seekg(2, ios::cur);
			in >> tmp.c.p[1].v[i];
			tmp.c.p[1].v[i]++;
		}
		in.get();
		ins.push_back(move(tmp));
	}
	return in;
}

static vector<Cuboid> split(const Cuboid& a, const Cuboid& b)
{
	vector<Cuboid> res;
	vector<Cuboid> tmp;

	res.push_back(a);
	for (int j = 0; j < 2; j++)
	{
		for (int i = 0; i < 3; i++)
		{
			auto& p = b.p[j];
			for (auto& c : res)
			{
				if (c.intersect(b)
				    && c.p[0].v[i] <= p.v[i]
				    && p.v[i] < c.p[1].v[i])
				{
					Cuboid n(c);
					n.p[1].v[i] = p.v[i];
					if (!n.inside(b))
					{
						tmp.push_back(n);
					}
					n = c;
					n.p[0].v[i] = p.v[i];
					if (!n.inside(b))
					{
						tmp.push_back(n);
					}
				}
				else
				{
					tmp.push_back(c);
				}
			}
			swap(res, tmp);
			tmp.clear();
		}
	}
	return res;
}

static void sub(vector<Cuboid>& lst, const Cuboid& b)
{
	vector<Cuboid> tmp;
	for (auto it = lst.begin(); it != lst.end(); )
	{
		if (it->inside(b))
		{
			it = lst.erase(it);
		}
		else if (!it->intersect(b))
		{
			++it;
		}
		else
		{
			auto r = split(*it, b);
			copy(r.begin(), r.end(), back_inserter(tmp));
			it = lst.erase(it);
		}
	}
	copy(tmp.begin(), tmp.end(), back_inserter(lst));
}

static void add(vector<Cuboid>& lst, const Cuboid& b)
{
	sub(lst, b);
	lst.push_back(b);
}

static uint64_t volume(const vector<Cuboid>& lst)
{
	return accumulate(lst.begin(), lst.end(), 0ULL, [](auto acc, auto &c){
		return acc + c.volume();
	});
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<Instr> prog;
	in >> prog;
	in.close();

	Cuboid region{{{-50,-50,-50},{51,51,51}}};
	uint64_t part1 = 0;
	vector<Cuboid> lst;
	for (auto& ins: prog)
	{
		if (!part1 && !ins.c.inside(region))
		{
			part1 = volume(lst);
		}
		switch (ins.cmd)
		{
		case Instr::ON: add(lst, ins.c); break;
		case Instr::OFF: sub(lst, ins.c); break;
		}
	}

	fmt::print("Part1: {}\n", part1);
	fmt::print("Part2: {}\n", volume(lst));

	return 0;
}

File: day22/day22-2.py
#!/usr/bin/env python3

class Cuboid(object):
    def __init__(self, p0, p1):
        self.p0 = p0
        self.p1 = p1

    def inside(self, other):
        for i in range(3):
            if self.p0[i] < other.p0[i] or self.p1[i] > other.p1[i]:
                return False
        return True

    def intersect(self, other):
        for i in range(3):
            if self.p1[i] <= other.p0[i] or self.p0[i] >= other.p1[i]:
                return False
        return True

    def volume(self):
        v = 1
        for i in range(3):
            v *= self.p1[i] - self.p0[i]
        return v

def parse_instruction(row):
    column = row.split()
    instr = column[0]
    coords = column[1].split(",")
    coords = [tuple(int(v) for v in x[2:].split("..")) for x in coords]

    (x0, x1), (y0, y1), (z0, z1) = coords
    return instr, Cuboid((x0, y0, z0), (x1+1, y1+1, z1+1))

def replace(p1, p2, i):
    p1 = list(p1)
    p1[i] = p2[i]
    return tuple(p1)

def split(c, b):
    r = []
    q = [c]
    for i in range(3):
        for c in q:
            if c.intersect(b) and c.p0[i] <= b.p0[i] < c.p1[i]:
                r.append(Cuboid(c.p0, replace(c.p1, b.p0, i)))
                r.append(Cuboid(replace(c.p0, b.p0, i), c.p1))
            else:
                r.append(c)
        q, r = r, []

        for c in q:
            if c.intersect(b) and c.p0[i] <= b.p1[i] < c.p1[i]:
                r.append(Cuboid(c.p0, replace(c.p1, b.p1, i)))
                r.append(Cuboid(replace(c.p0, b.p1, i), c.p1))
            else:
                r.append(c)
        q, r = r, []

    return q

def sub(lst, b):
    r = []
    for c in lst:
        if not c.intersect(b):
            r.append(c)
        else:
            for a in split(c, b):
                if not a.inside(b):
                    r.append(a)
    return r

def add(lst, b):
    lst = sub(lst, b)
    lst.append(b)
    return lst

def volume(lst):
    return sum(c.volume() for c in lst)

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        region = Cuboid((-50, -50, -50), (51, 51, 51))
        part1 = 0
        lst = []
        for ins, cuboid in (parse_instruction(row) for row in txt.splitlines()):
            if not part1 and not cuboid.inside(region):
                part1 = volume(lst)
            if ins == "on":
                lst = add(lst, cuboid)
            elif ins == "off":
                lst = sub(lst, cuboid)

        print("Part1:", part1)
        print("Part2:", volume(lst))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
